doc/build-helpers/fetchers.chapter.md

```diff
@@ -1,88 +1,540 @@
-};
+  url = "https://raw.githubusercontent.com/NixOS/nixpkgs/23.05/.version";
+  hash = "sha256-ZHl1emidXVojm83LCVrwULpwIzKE/mYwfztVkvpruOM=";
+}
 ```
 
-A common mistake is to update a fetcher’s URL, or a version parameter, without updating the hash.
+A common mistake is to update a fetcher parameter, such as `url`, without updating the hash:
 
 ```nix
+{ fetchurl }:
 fetchurl {
-  url = "http://www.example.org/hello-1.1.tar.gz";
-  hash = "sha256-lTeyxzJNQeMdu1IVdovNMtgn77jRIhSybLdMbTkf2Ww=";
-};
+  url = "https://raw.githubusercontent.com/NixOS/nixpkgs/23.11/.version";
+  hash = "sha256-ZHl1emidXVojm83LCVrwULpwIzKE/mYwfztVkvpruOM=";
+}
 ```
 
-**This will reuse the old contents**.
-Remember to invalidate the hash argument, in this case by setting the `hash` attribute to an empty string.
+**This will produce the same output as before!**.
 
 ```nix
+{ fetchurl }:
 fetchurl {
-  url = "http://www.example.org/hello-1.1.tar.gz";
+  url = "https://raw.githubusercontent.com/NixOS/nixpkgs/23.11/.version";
   hash = "";
-};
+}
 ```
 
-Use the resulting error message to determine the correct hash.
+When building the package, use the error message to determine the correct hash:
 
-```
-error: hash mismatch in fixed-output derivation '/path/to/my.drv':
+```shell
+$ nix-build
+...
+error: hash mismatch in fixed-output derivation '/nix/store/7yynn53jpc93l76z9zdjj4xdxgynawcw-version.drv':
          specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
-            got:    sha256-lTeyxzJNQeMdu1IVdovNMtgn77jRIhSybLdMbTkf2Ww=
+            got:    sha256-BZqI7r0MNP29yGH5+yW2tjU9OOpOCEvwWKrWCv5CQ0I=
+error: build of '/nix/store/bqdjcw5ij5ymfbm41dq230chk9hdhqff-version.drv' failed
+```
+:::
+
+## Obtaining hashes securely {#sec-pkgs-fetchers-secure-hashes}
+
+Let's say Man-in-the-Middle (MITM) sits close to your network.
+Then instead of fetching source you can fetch malware, and instead of the actual source hash, you get the hash of malware.
+Here are security considerations for this scenario:
+
+- `http://` URLs are not secure to prefetch hashes.
+
+- Upstream hashes should be obtained via a secure protocol.
+
+- `https://` URLs are secure when using `nix-prefetch-*` or for upstream hashes.
+
+- `https://` URLs are secure relying on "trust on first use" *only if* you use one of the listed fake hashes.
+  If you use any other hash, `fetchurl` will pass `--insecure` to `curl` and may then degrade to HTTP in case of TLS certificate expiration.
+
+[]{#fetchurl}
+## `fetchurl` {#sec-pkgs-fetchers-fetchurl}
+
+`fetchurl` returns a derivation which downloads content from a given URL and stores the unaltered contents within the Nix store.
+
+It uses {manpage}`curl(1)` internally, and allows its behaviour to be modified by specifying a few attributes in the argument to `fetchurl` (see the documentation for attributes `curlOpts`, `curlOptsList`, and `netrcPhase`).
+
+### File naming {#sec-pkgs-fetchers-fetchurl-file-naming}
```

Comment:
```suggestion
```

Don't think this is needed. The section is not that long.

Reply:
Thanks! I split this PR into two, and I applied this suggestion in #300429 which is the PR with the fetchurl changes.